R/implicit_assignment_linter.R

```diff
@@ -0,0 +1,90 @@
+#' Avoid implicit assignment in function calls
+#'
+#' Assigning inside function calls makes the code difficult to read, and should
+#' be avoided, except for functions that capture side-effects (e.g. [capture.output()]).
+#'
+#' @examples
+#' # will produce lints
+#' lint(
+#'   text = "if (x <- 1L) TRUE",
+#'   linters = implicit_assignment_linter()
+#' )
+#'
+#' lint(
+#'   text = "mean(x <- 1:4)",
+#'   linters = implicit_assignment_linter()
+#' )
+#'
+#' # okay
+#' writeLines("x <- 1L\nif (x) TRUE")
+#' lint(
+#'   text = "x <- 1L\nif (x) TRUE",
+#'   linters = implicit_assignment_linter()
+#' )
+#'
+#' writeLines("x <- 1:4\nmean(x)")
+#' lint(
+#'   text = "x <- 1:4\nmean(x)",
+#'   linters = implicit_assignment_linter()
+#' )
+#'
+#' @evalRd rd_tags("implicit_assignment_linter")
+#' @seealso
+#' - [linters] for a complete list of linters available in lintr.
+#' - <https://style.tidyverse.org/syntax.html#assignment>
+#'
+#' @export
+implicit_assignment_linter <- function() {
+  assignments <- c(
+    "LEFT_ASSIGN", # e.g. mean(x <- 1:4)
+    "RIGHT_ASSIGN" # e.g. mean(1:4 -> x)
+  )
+  xpath_fun_call <- paste0(
+    "
+    //SYMBOL_FUNCTION_CALL[not(text() = 'capture.output')]
```

Comment:
These exceptions should be configurable.
I can see `test_that()`, `local()` and language functions like `quote()` joining that list by default, but there are probably others.